orttraining/orttraining/core/graph/training_op_defs.cc

```diff
@@ -1265,6 +1263,90 @@ void RegisterTrainingOpSchemas() {
           {"tensor(bool)"},
           "Constrain types to boolean tensors.");
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(Adam)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Input(0, "lr", "The initial learning rate.", "T1")
+      .Input(1, "step", "The update count of \"X\". It should be a scalar.", "T2")
+      .Input(2, "weights", "Sequence of weights to optimize.", "S_WEIGHT")
+      .Input(3, "gradients", "Sequence of gradients computed in this iteration.", "S_GRAD")
+      .Input(4, "momentums_1", "Sequence of exponentially averaged historical gradients.", "S_MOMENT")
+      .Input(5, "momentums_2", "Sequence of exponentially averaged historical squared gradients.", "S_MOMENT")
+      .Output(0, "updated_flag", "Whether gradient is applied or not.", "T2")
```

Comment:
Currently never. There are four outputs, the last three are optional, the first one is required. The reason to do this, because AdamWOptimizer op will be likely produce graph outputs, and the operator itself most of the time won't need produce outputs because it is trying to optimize weights and momentums inplace, both as the node's inputs. 

So for real training, we just need pull the updated_flag, which is a trigger for the graph run, no other meaning (imagine once data goes into optimizer, it will be applied anyway). 
The reason we add other 3 outputs are for testing purpose currently. 

The same things could be applied to InplaceAccumulation op, while I saw from Aish's PR, she is adding a special branch handle that op in the allocation planner, to skip the graph outputs copy.